src/xrpld/rpc/handlers/LedgerEntry.cpp

```diff
@@ -66,768 +67,819 @@ parseAuthorizeCredentials(Json::Value const& jv)
-                    else
-                        uNodeIndex =
-                            keylet::mptoken(mptIssuanceID, *account).key;
-                }
-                catch (std::runtime_error const&)
-                {
-                    jvResult[jss::error] = "malformedRequest";
-                }
-            }
+            jvResult[jss::error] = "malformedAddress";
+            return std::nullopt;
         }
-        else
+
+        return keylet::mptoken(mptIssuanceID, *account).key;
+    }
+    catch (std::runtime_error const&)
+    {
+        jvResult[jss::error] = "malformedRequest";
+        return std::nullopt;
+    }
+}
+
+using FunctionType =
+    std::function<std::optional<uint256>(Json::Value const&, Json::Value&)>;
+
+struct LedgerEntry
+{
+    Json::StaticString fieldName;
+    FunctionType parseFunction;
+    LedgerEntryType expectedType;
+};
+
+// {
+//   ledger_hash : <ledger>
+//   ledger_index : <ledger_index>
+//   ...
+// }
+Json::Value
+doLedgerEntry(RPC::JsonContext& context)
+{
+    std::shared_ptr<ReadView const> lpLedger;
+    auto jvResult = RPC::lookupLedger(lpLedger, context);
+
+    if (!lpLedger)
+        return jvResult;
+
+    static LedgerEntry ledgerEntryParsers[] = {
+        {jss::index, parseIndex, ltANY},
+        {jss::account_root, parseAccountRoot, ltACCOUNT_ROOT},
+        // TODO: add amendments
+        {jss::amm, parseAMM, ltAMM},
+        {jss::bridge, parseBridge, ltBRIDGE},
+        {jss::check, parseCheck, ltCHECK},
+        {jss::credential, parseCredential, ltCREDENTIAL},
+        {jss::deposit_preauth, parseDepositPreauth, ltDEPOSIT_PREAUTH},
+        {jss::did, parseDID, ltDID},
+        {jss::directory, parseDirectory, ltDIR_NODE},
+        {jss::escrow, parseEscrow, ltESCROW},
+        // TODO: add fee, hashes
+        {jss::mpt_issuance, parseMPTokenIssuance, ltMPTOKEN_ISSUANCE},
+        {jss::mptoken, parseMPToken, ltMPTOKEN},
+        // TODO: add NFT Offers
+        {jss::nft_page, parseNFTokenPage, ltNFTOKEN_PAGE},
+        // TODO: add NegativeUNL
+        {jss::offer, parseOffer, ltOFFER},
+        {jss::oracle, parseOracle, ltORACLE},
+        {jss::payment_channel, parsePaymentChannel, ltPAYCHAN},
+        {jss::ripple_state, parseRippleState, ltRIPPLE_STATE},
+        // This is an alias, since the `ledger_data` filter uses jss::state
+        {jss::state, parseRippleState, ltRIPPLE_STATE},
+        {jss::ticket, parseTicket, ltTICKET},
+        {jss::xchain_owned_claim_id,
+         parseXChainOwnedClaimID,
+         ltXCHAIN_OWNED_CLAIM_ID},
+        {jss::xchain_owned_create_account_claim_id,
+         parseXChainOwnedCreateAccountClaimID,
+         ltXCHAIN_OWNED_CREATE_ACCOUNT_CLAIM_ID},
+    };
+
+    uint256 uNodeIndex;
+    LedgerEntryType expectedType = ltANY;
+
+    try
+    {
+        bool found = false;
+        for (const auto& ledgerEntry : ledgerEntryParsers)
         {
-            if (context.params.isMember("params") &&
```

Comment:
This was only used for the CLI, so the CLI logic was moved to `RPCCall` instead, where it's cleaner and easier to read.